### ISSUE TITLE
Test failures on Python 3.8

### ISSUE THREAD

**Issue by hroncok** [spulec/freezegun#293]
Hi. I'm bootstrapping Python 3.8 in Fedora and I got stuck at freezegun, it has build fialures. Those are reproducible here, on master.

```
$ git rev-parse HEAD
9347d133f33f675c87bb0569d70d9d95abef737f
$ tox -e py38
GLOB sdist-make: .../freezegun/setup.py
py38 inst-nodeps: .../freezegun/.tox/dist/freezegun-0.3.11.zip
py38 installed: atomicwrites==1.3.0,attrs==19.1.0,certifi==2019.3.9,chardet==3.0.4,coverage==4.5.3,coveralls==1.7.0,dateparser==0.7.1,docopt==0.6.2,freezegun==0.3.11,humanize==0.5.1,idna==2.8,maya==0.6.1,mock==2.0.0,more-itertools==6.0.0,pbr==5.1.3,pendulum==2.0.4,pluggy==0.9.0,py==1.8.0,pytest==3.10.1,pytest-cov==2.6.1,pytest-cover==3.0.0,pytest-coverage==0.0,python-dateutil==2.8.0,pytz==2018.9,pytzdata==2018.9,regex==2019.3.12,requests==2.21.0,six==1.12.0,snaptime==0.2.4,tzlocal==1.5.1,urllib3==1.24.1
py38 run-test-pre: PYTHONHASHSEED='905734949'
py38 runtests: commands[0] | make test NOSE_ARGS=
rm -f .coverage
pytest
============================= test session starts ==============================
platform linux -- Python 3.8.0a2, pytest-3.10.1, py-1.8.0, pluggy-0.9.0
rootdir: .../freezegun, inifile:
plugins: cov-2.6.1
collected 105 items

tests/test_asyncio.py ..                                                 [  1%]
tests/test_class_import.py ..............                                [ 15%]
tests/test_datetimes.py ..................................s............. [ 60%]
..........                                                               [ 70%]
tests/test_errors.py ..                                                  [ 72%]
tests/test_import_alias.py ...                                           [ 75%]
tests/test_operations.py .........                                       [ 83%]
tests/test_pickle.py ...                                                 [ 86%]
tests/test_sqlite3.py ..                                                 [ 88%]
tests/test_ticking.py ......                                             [ 94%]
tests/test_utils.py ..                                                   [ 96%]
tests/test_uuid.py FF                                                    [ 98%]
tests/test_warnings.py ..                                                [100%]

=================================== FAILURES ===================================
______________________________ test_uuid1_future _______________________________

    def test_uuid1_future():
        """
        Test that we can go back in time after setting a future date.
        Normally UUID1 would disallow this, since it keeps track of
        the _last_timestamp, but we override that now.
        """
        future_target = datetime.datetime(2056, 2, 6, 14, 3, 21)
        with freeze_time(future_target):
>           assert time_from_uuid(uuid.uuid1()) == future_target
E           AssertionError: assert FakeDatetime(2019, 3, 28, 9, 59, 29, 624778) == datetime.datetime(2056, 2, 6, 14, 3, 21)
E            +  where FakeDatetime(2019, 3, 28, 9, 59, 29, 624778) = time_from_uuid(UUID('2d98abed-5140-11e9-af9c-e86a64040576'))
E            +    where UUID('2d98abed-5140-11e9-af9c-e86a64040576') = <function uuid1 at 0x7fe23a36b5c0>()
E            +      where <function uuid1 at 0x7fe23a36b5c0> = uuid.uuid1

tests/test_uuid.py:26: AssertionError
_______________________________ test_uuid1_past ________________________________

    def test_uuid1_past():
        """
        Test that we can go forward in time after setting some time in the past.
        This is simply the opposite of test_uuid1_future()
        """
        past_target = datetime.datetime(1978, 7, 6, 23, 6, 31)
        with freeze_time(past_target):
>           assert time_from_uuid(uuid.uuid1()) == past_target
E           AssertionError: assert FakeDatetime(2019, 3, 28, 9, 59, 29, 664840) == datetime.datetime(1978, 7, 6, 23, 6, 31)
E            +  where FakeDatetime(2019, 3, 28, 9, 59, 29, 664840) = time_from_uuid(UUID('2d9ec8d5-5140-11e9-b7ca-e86a64040576'))
E            +    where UUID('2d9ec8d5-5140-11e9-b7ca-e86a64040576') = <function uuid1 at 0x7fe23a36b5c0>()
E            +      where <function uuid1 at 0x7fe23a36b5c0> = uuid.uuid1

tests/test_uuid.py:40: AssertionError
=============================== warnings summary ===============================
.tox/py38/lib/python3.8/site-packages/dateparser/date.py:320
  .../freezegun/.tox/py38/lib/python3.8/site-packages/dateparser/date.py:320: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(languages, (list, tuple, collections.Set)) and languages is not None:

tests/test_datetimes.py::test_should_use_real_time
  .../freezegun/freezegun/api.py:207: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    return real_clock()

tests/test_ticking.py::test_ticking_time_clock
  .../freezegun/freezegun/api.py:210: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    return 0.0 if not tick_flags[-1] else real_clock()
  .../freezegun/freezegun/api.py:219: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    total_seconds += real_clock()
  .../freezegun/freezegun/api.py:219: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    total_seconds += real_clock()
  .../freezegun/freezegun/api.py:219: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    total_seconds += real_clock()
  .../freezegun/freezegun/api.py:219: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
    total_seconds += real_clock()

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 102 passed, 1 skipped, 7 warnings in 0.74 seconds ==========
make: *** [Makefile:9: test] Error 1
ERROR: InvocationError for command '/usr/bin/make test NOSE_ARGS=' (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   py38: commands failed
```

There are also some Python 3.8 deprecation warnings.

**Comment by tirkarthi** [comment]
`time.clock` is now removed in 3.8 : https://github.com/python/cpython/pull/13270/

**Comment by pganssle** [comment]
PR #296 fixes the `time.clock` issue, but not the UUID problem.

**Comment by cooperlees** [comment]
Also getting these failures with 3.8 with bandersnatch's tests (https://github.com/pypa/bandersnatch). Just signaling and will try work on a fix next time I get some spare time.

Example Failure: https://travis-ci.org/pypa/bandersnatch/jobs/547222192

**Comment by tirkarthi** [comment]
I did some bisecting and reverting https://github.com/python/cpython/commit/62a68b762a479a72c3defba9ace5f72a0063c5c6 seems to fix the test failure in freezetime regarding uuid. The change uses `time.time_ns` instead of `time.time` which perhaps was not handled in freezegun while mocking?

```python
# /tmp/foo.py
import datetime
import uuid
import freezegun

def time_from_uuid(value):
     uvalue = value if isinstance(value, uuid.UUID) else uuid.UUID(value)
     assert uvalue.version == 1
     return (datetime.datetime(1582, 10, 15) +
             datetime.timedelta(microseconds=uvalue.time // 10))

past_target = datetime.datetime(1978, 7, 6, 23, 6, 31)

with freezegun.freeze_time(past_target):
    print(time_from_uuid(uuid.uuid1()))
```

```
(freezegun-venv) karthi@ubuntu-s-1vcpu-1gb-blr1-01:~/cpython$ python -W ignore /tmp/foo.py
2019-06-18 17:44:19.170758
(freezegun-venv) karthi@ubuntu-s-1vcpu-1gb-blr1-01:~/cpython$ git checkout 62a68b762a479a72c3defba9ace5f72a0063c5c6~1 Lib/uuid.py
(freezegun-venv) karthi@ubuntu-s-1vcpu-1gb-blr1-01:~/cpython$ python -W ignore /tmp/foo.py
1978-07-06 23:06:31
```

**Comment by tirkarthi** [comment]
I tried patching `time.time_ns` like other time related functions and the test passes. `time.time_ns` is from 3.7 only so there should be some version check to include the code only for 3.7+ . A simple patch as below. I will try to convert this into a PR.

```patch
diff --git a/freezegun/api.py b/freezegun/api.py
index 5e7d7fa..ecbf5f2 100644
--- a/freezegun/api.py
+++ b/freezegun/api.py
@@ -22,12 +22,13 @@ except ImportError:


 real_time = time.time
+real_ns_time = time.time_ns
 real_localtime = time.localtime
 real_gmtime = time.gmtime
 real_strftime = time.strftime
 real_date = datetime.date
 real_datetime = datetime.datetime
-real_date_objects = [real_time, real_localtime, real_gmtime, real_strftime, real_date, real_datetime]
+real_date_objects = [real_time, real_ns_time, real_localtime, real_gmtime, real_strftime, real_date, real_datetime]
 _real_time_object_ids = set(id(obj) for obj in real_date_objects)

 # time.clock is deprecated and was removed in Python 3.8
@@ -176,6 +177,12 @@ def fake_time():
     return calendar.timegm(current_time.timetuple()) + current_time.microsecond / 1000000.0


+def fake_ns_time():
+    if _should_use_real_time():
+        return real_time_ns()
+    return int(int(fake_time()) * 1e9)
+
+
 def fake_localtime(t=None):
     if t is not None:
         return real_localtime(t)
@@ -578,6 +585,7 @@ class _freeze_time(object):
         datetime.date = FakeDate

         time.time = fake_time
+        time.time_ns = fake_ns_time
         time.localtime = fake_localtime
         time.gmtime = fake_gmtime
         time.strftime = fake_strftime
@@ -597,6 +605,7 @@ class _freeze_time(object):
             ('real_localtime', real_localtime, fake_localtime),
             ('real_strftime', real_strftime, fake_strftime),
             ('real_time', real_time, fake_time),
+            ('real_ns_time', real_ns_time, fake_ns_time),
         ]

         if real_clock is not None:
```

**Comment by pganssle** [comment]
@tirkarthi How about something like:

```python
_NEEDS_TIME_NS = getattr(time, 'time_ns', None) is not None
```

Then gate the new mocks on that?

**Comment by tirkarthi** [comment]
@pganssle Thanks for the suggestion. It's cleaner than using try/except. I will use `hasattr` and tests already use this pattern : https://github.com/spulec/freezegun/blob/64a0541cec8e9e3d7b63fdccb80db19cd7a613fe/tests/test_datetimes.py#L21. I created PR #305 . I was not able to group all `time.time_ns` under one place which would be nice.

**Comment by tirkarthi** [comment]
This should be fixed now on master and 3.8-dev is green on CI. There are some additional deprecation warnings regarding asyncio covered as part of #303 .

**Comment by FedericoCeratto** [comment]
@spulec Python 3.8 is gaining popularity. Is there any plan to release this fix?

**Comment by spulec** [comment]
I've released version 0.3.13 now!
